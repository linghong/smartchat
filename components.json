{
  "$schema": "https://ui.shadcn.com/schema.json",
  "style": "default",
  "rsc": true,
  "tsx": true,
  "tailwind": {
    "config": "tailwind.config.cjs",
    "css": "src/styles/globals.css",
    "baseColor": "gray",
    "cssVariables": false,
    "prefix": ""
  },
  "aliases": {
    "components": "@/src/components",
    "utils": "@/src/lib/utils",
    "ui": "@/src/components/ui",
    "lib": "@/src/utils/lib",
    "hooks": "@/src/hooks"
  }
}
