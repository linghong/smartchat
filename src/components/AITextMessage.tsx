import CodeBlock from '@/src/components/CodeBlock';
import { marked } from 'marked';

/**
 * Function to process a text message
 * This functions do the following things:
 * convert new line to <br>
 * add class="inline-code" in <code /> tag
 * - Adds <pre><code class="language-xxx"></code></pre> tags around code blocks with language
 * @param text - The string.
 * @returns The processed string.
 */
export const format = (text: string) => {
  // First handle the case where <br> tags are already present, for example, gpt4-o's message has <br>
  let formatted = text.replace(/<br>/g, '\n');

  // Protect code blocks with language specifier by replacing newlines within them
  formatted = formatted.replace(
    /```(\w+)?\n([\s\S]*?)```/g,
    (match, lang, code) => {
      const protectedCode = code.replace(/\n/g, '&!newline!&');
      return `\`\`\`${lang || ''}&!newline!&${protectedCode}&!newline!&\`\`\``;
    }
  );

  // Then, replace newlines with <br> tags
  formatted = formatted.replace(/\n/g, '<br>');

  // Add a consistent class to inline <code> tags
  formatted = formatted.replace(
    /<code>(.*?)<\/code>/g,
    (match, p1) => `<code class="inline-code">${p1}</code>`
  );

  // Restore the newlines within code blocks and add appropriate tags
  formatted = formatted.replace(
    /```(\w+)?&!newline!&([\s\S]*?)&!newline!&```/g,
    (match, lang, code) => {
      const cleanCode = code.replace(/&!newline!&/g, '\n').trim();
      const languageClass = lang ? `class="language-${lang}"` : '';
      return `<pre><code ${languageClass}>${cleanCode}</code></pre>`;
    }
  );

  return formatted;
};

const AITextMessage: React.FC<{ content: string }> = ({ content }) => {
  const formattedContent = format(content);

  // separate text and code
  const parts = [];

  const codeRegex =
    /<pre><code(?: class="language-(\w+)")?>([\s\S]*?)<\/code><\/pre>/g;
  let lastIndex = 0;
  let match;

  while ((match = codeRegex.exec(formattedContent)) !== null) {
    if (match.index > lastIndex) {
      parts.push(
        <span
          key={lastIndex}
          dangerouslySetInnerHTML={{
            __html: marked(
              formattedContent.slice(lastIndex, match.index)
            ) as string
          }}
        />
      );
    }

    const [, language, code] = match;
    parts.push(<CodeBlock key={match.index} code={code} language={language} />);

    lastIndex = match.index + match[0].length;
  }

  // When there is zero codeblock in the text
  // Or when there is no more codeblocks found, but still have some text left
  if (lastIndex < formattedContent.length) {
    parts.push(
      <span
        key={lastIndex}
        dangerouslySetInnerHTML={{
          __html: marked(formattedContent.slice(lastIndex)) as string
        }}
      />
    );
  }

  // return an array of components
  return <>{parts}</>;
};

export default AITextMessage;
