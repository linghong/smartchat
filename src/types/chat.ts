export interface Chat {
  id: number;
  title: string;
}

export type ChatRole = 'system' | 'assistant' | 'user';

export type ChatType = {
  role: ChatRole;
  content: string;
};

export interface ImageFile {
  base64Image: string;
  mimeType: string;
  size: number;
  name: string;
  width?: number;
  height?: number;
}

export interface FileData {
  base64Content: string;
  type: string;
  size: number;
  name: string;
}

export interface Message {
  question: string;
  answer: string;
  model: string;
}

export interface OpenAIChatContentImage {
  type: 'image_url';
  image_url: {
    url: string;
    detail: 'auto' | 'low' | 'high';
  };
}
export interface OpenAIChatContentText {
  type: 'text';
  text: string;
}
