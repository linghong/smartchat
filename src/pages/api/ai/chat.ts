import type { NextApiRequest, NextApiResponse } from 'next';
import { marked } from 'marked';

import { fetchDataFromPinecone } from '@/src/services/fetchDataFromPinecone';

import getClaudeChatCompletion from '@/src/services/llm/claude';
import getGeminiChatCompletion from '@/src/services/llm/gemini';
import { getGroqChatCompletion } from '@/src/services/llm/groq';
import {
  createEmbedding,
  getOpenAIChatCompletion
} from '@/src/services/llm/openai';
import getOpenModelChatCompletion from '@/src/services/llm/opensourceai';

import {
  processImageFiles,
  processNonMediaFiles
} from '@/src/utils/processMessageFile';
import { FileData, ImageFile } from '@/src/types/chat';
import {
  extractMessageContent,
  extractSubjectTitle
} from '@/src/utils/chatMessageHelper';

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  const { basePrompt, namespace, selectedModel, chatHistory, fileSrc } =
    req.body;

  let { question } = req.body;

  if (!question && fileSrc.length === 0) {
    return res.status(400).json({ message: 'No question in the request' });
  }

  if (!selectedModel?.value) {
    return res.status(500).json('Model name is missing.');
  }

  const base64ImageSrc = processImageFiles(fileSrc);

  question += await processNonMediaFiles(fileSrc);

  // Replace multiple newlines with a single newline to avoid excessive spacing
  const pretreatedQuestion = question.trim().replace(/\n\s*\n/g, '\n');

  try {
    let fetchedText = '';

    //fetch text from vector database
    if (namespace && namespace !== 'none') {
      const embeddedQuery = await createEmbedding(question);
      fetchedText = await fetchDataFromPinecone(embeddedQuery, namespace);
    }

    // get response from AI
    const { category } = selectedModel;
    let chatResponse: string | undefined | null;

    switch (category) {
      case 'openai':
        chatResponse = await getOpenAIChatCompletion(
          basePrompt,
          chatHistory,
          pretreatedQuestion,
          fetchedText,
          selectedModel,
          base64ImageSrc
        );
        break;

      case 'groq':
        chatResponse = await getGroqChatCompletion(
          basePrompt,
          chatHistory,
          pretreatedQuestion,
          fetchedText,
          selectedModel
        );
        break;

      case 'google':
        chatResponse = await getGeminiChatCompletion(
          basePrompt,
          chatHistory,
          pretreatedQuestion,
          fetchedText,
          selectedModel,
          base64ImageSrc
        );
        break;

      case 'anthropic':
        chatResponse = await getClaudeChatCompletion(
          basePrompt,
          chatHistory,
          pretreatedQuestion,
          fetchedText,
          selectedModel,
          base64ImageSrc
        );
        break;

      case 'hf-small':
      case 'hf-large':
        const baseUrl =
          category === 'hf-small'
            ? process.env.NEXT_PUBLIC_SERVER_URL
            : process.env.NEXT_PUBLIC_SERVER_GPU_URL;
        if (!baseUrl) {
          return res
            .status(500)
            .json(
              `Url address for posting the data to ${selectedModel} is missing`
            );
        }
        const url = `${baseUrl}/api/chat_${category === 'hf-small' ? 'cpu' : 'gpu'}`;
        chatResponse = await getOpenModelChatCompletion(
          basePrompt,
          chatHistory,
          pretreatedQuestion,
          fetchedText,
          selectedModel,
          url
        );
        break;

      default:
        return res.status(500).json('Invalid model category');
    }

    let answer: string;
    let subject: string;
    if (!chatResponse) {
      answer = "Sorry, I'm having trouble finding an answer to your question.";
      subject = 'Unknown';
    } else {
      answer = extractMessageContent(chatResponse);
      subject = extractSubjectTitle(chatResponse);
    }

    const htmlAnswer = marked(answer);

    res.status(200).json({ answer: htmlAnswer, subject });
  } catch (error: any) {
    console.error('An error occurred: ', error);
    res.status(500).json({ error: error.message || 'Something went wrong' });
  }
}
