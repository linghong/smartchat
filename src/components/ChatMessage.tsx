import Image from 'next/image'

import { Message, ImageFile } from '@/src/types/chat'
import ImageListWithModal from '@/src/components/ImageListWithModal'

type ChatMessageProps = {
  index: number
  message: Message
  lastIndex: boolean
  loading: boolean
  imageSrc: ImageFile[]
  modelName: string
  handleImageDelete: (e: any) => void
}

const ChatMessage: React.FC<ChatMessageProps> = ({
  index,
  message,
  imageSrc,
  lastIndex,
  modelName,
  loading,
  handleImageDelete,
}) => {
  const convertNewlinesToBreaks = (text: string) => {
    return text.replace(/\n/g, '<br>')
  }

  return (
    <>
      {index !== 0 && (
        <article
          className="flex px-3 py-2 text-black"
          aria-label="user-message"
        >
          <div className="w-16 flex flex-col justify-start items-center ai-answer">
            <Image
              src="/user.png"
              alt="User avatar"
              width="30"
              height="30"
              className="h-8 w-8 mr-4 rounded-sm"
              priority
            />
          </div>
          <div className="w-11/12 user-question">
            <div
              className="space-wrap"
              dangerouslySetInnerHTML={{
                __html: convertNewlinesToBreaks(message.question),
              }}
            />
            {
              <ImageListWithModal
                imageSrc={imageSrc}
                handleImageDelete={handleImageDelete}
              />
            }
          </div>
        </article>
      )}
      {
        <article
          className="flex px-3 py-2 bg-slate-100 color-black ease-in duration-300"
          aria-labelledby="ai-response"
        >
          <div className="w-16 flex flex-col justify-start items-center ai-answer">
            <Image
              src="/bot.png"
              alt="AI bot avatar"
              width="30"
              height="30"
              className={`h-8 w-8 mr-4 rounded-sm ${loading && lastIndex && 'animate-pulse'}`}
              priority
            />
            <label className="text-xs">{modelName}</label>
          </div>
          <div
            className=" bot-response answer space-wrap"
            dangerouslySetInnerHTML={{
              __html: convertNewlinesToBreaks(message.answer),
            }}
          />
        </article>
      }
    </>
  )
}

export default ChatMessage
