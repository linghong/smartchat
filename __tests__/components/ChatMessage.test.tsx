import React from 'react';
import { render, screen, act, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

import ChatMessage from '@/src/components/ChatMessage';
import { format } from '@/src/components/AITextMessage';
import { encodeHTMLEntities } from '@/src/utils/guardrails/htmlEncodeDecode';
import { Message, FileData, AssistantOption } from '@/src/types/chat';

// Mock the console methods to suppress output during tests
beforeAll(() => {
  jest.spyOn(console, 'log').mockImplementation(() => {});
  jest.spyOn(console, 'warn').mockImplementation(() => {});
  jest.spyOn(console, 'error').mockImplementation(() => {});
});

afterAll(() => {
  jest.restoreAllMocks();
});

describe('ChatMessage Component', () => {
  let handleFileDelete: jest.Mock;
  let handleAssistantChange: jest.Mock;
  let handleRetry: jest.Mock;
  let handleCopy: jest.Mock;

  const message: Message = {
    question: 'What is AI?',
    answer: 'AI stands for Artificial Intelligence.',
    assistant: 'gpt-4'
  };
  const loading = false;
  const lastIndex = true;

  const selectedAssistant: AssistantOption = {
    isDefault: false,
    value: 'GPT-4-assistant',
    label: 'GPT-4 Assistant',
    config: {
      name: 'GPT-4-assistant',
      role: 'coding',
      model: {
        value: 'gpt-4',
        label: 'GPT-4',
        category: 'openai',
        contextWindow: 128000,
        vision: true
      },
      basePrompt: '',
      temperature: 0.1,
      topP: 1.0
    }
  };
  const assistantOptions: AssistantOption[] = [
    {
      isDefault: true,
      value: 'default-gpt-3.5-turbo',
      label: 'Default GPT-3.5 Turbo',
      config: {
        name: 'Default GPT-3.5 Turbo',
        role: 'writing',
        model: {
          value: 'gpt-3.5-turbo',
          label: 'GPT-3.5 Turbo',
          category: 'openai',
          contextWindow: 128000,
          vision: false
        },
        basePrompt: '',
        temperature: 0.4,
        topP: 0.7
      }
    },
    selectedAssistant
  ];

  const fileSrc: FileData[] = [
    {
      base64Content: '/path/to/image1.png',
      type: 'image/png',
      size: 5000,
      name: 'image1'
    },
    {
      base64Content: 'path/to/image2.png',
      type: 'image/png',
      size: 8000,
      name: 'image2'
    }
  ];

  beforeEach(() => {
    handleFileDelete = jest.fn();
    handleAssistantChange = jest.fn();
    handleRetry = jest.fn();
    handleCopy = jest.fn();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders ChatMessage component correctly with question and answer', () => {
    render(
      <ChatMessage
        isNew={false}
        message={message}
        lastIndex={lastIndex}
        loading={loading}
        fileSrc={fileSrc}
        selectedAssistant={selectedAssistant}
        handleAssistantChange={handleAssistantChange}
        assistantOptions={assistantOptions}
        handleFileDelete={handleFileDelete}
        handleRetry={handleRetry}
        handleCopy={handleCopy}
      />
    );
    expect(screen.getByText('What is AI?')).toBeInTheDocument();
    expect(
      screen.getByText('AI stands for Artificial Intelligence.')
    ).toBeInTheDocument();
  });

  it('renders the user and AI avatars', () => {
    render(
      <ChatMessage
        isNew={false}
        message={message}
        lastIndex={lastIndex}
        loading={loading}
        fileSrc={fileSrc}
        selectedAssistant={selectedAssistant}
        handleAssistantChange={handleAssistantChange}
        assistantOptions={assistantOptions}
        handleFileDelete={handleFileDelete}
        handleRetry={handleRetry}
        handleCopy={handleCopy}
      />
    );
    expect(screen.getByAltText('User avatar')).toBeInTheDocument();
    expect(screen.getByAltText('AI avatar')).toBeInTheDocument();
  });

  it('applies animate-pulse class when loading and is the last index', () => {
    act(() => {
      render(
        <ChatMessage
          isNew={false}
          message={message}
          lastIndex={true}
          loading={true}
          fileSrc={fileSrc}
          selectedAssistant={selectedAssistant}
          handleAssistantChange={handleAssistantChange}
          assistantOptions={assistantOptions}
          handleFileDelete={handleFileDelete}
          handleRetry={handleRetry}
          handleCopy={handleCopy}
        />
      );
    });
    const botImages = screen.getAllByAltText('AI avatar') as HTMLElement[];
    expect(botImages[botImages.length - 1]).toHaveClass('animate-pulse');
  });

  it('does not apply animate-pulse class when not loading', () => {
    act(() => {
      render(
        <ChatMessage
          isNew={true}
          message={message}
          lastIndex={true}
          loading={false}
          fileSrc={fileSrc}
          selectedAssistant={selectedAssistant}
          handleAssistantChange={handleAssistantChange}
          assistantOptions={assistantOptions}
          handleFileDelete={handleFileDelete}
          handleRetry={handleRetry}
          handleCopy={handleCopy}
        />
      );
    });

    const botImage = screen.getByAltText('AI avatar');
    expect(botImage).not.toHaveClass('animate-pulse');
  });

  it('does not render user message when isNew is true', () => {
    render(
      <ChatMessage
        isNew={true}
        message={message}
        lastIndex={true}
        loading={false}
        fileSrc={fileSrc}
        selectedAssistant={selectedAssistant}
        handleAssistantChange={handleAssistantChange}
        assistantOptions={assistantOptions}
        handleFileDelete={handleFileDelete}
        handleRetry={handleRetry}
        handleCopy={handleCopy}
      />
    );
    expect(screen.queryByText('What is AI?')).not.toBeInTheDocument();
  });

  it('sanitizes and formats messages correctly', () => {
    const userMessage = `Is <script>alert("XSS")</script> sanitized? 
    it is <strong>sanitized</strong>`;
    const messageWithHtml = {
      ...message,
      question: userMessage,
      answer:
        'Yes, it is <strong>sanitized</strong> and <code>formatted</code>.'
    };
    render(
      <ChatMessage
        isNew={false}
        message={messageWithHtml}
        lastIndex={lastIndex}
        loading={loading}
        fileSrc={fileSrc}
        selectedAssistant={selectedAssistant}
        handleAssistantChange={handleAssistantChange}
        assistantOptions={assistantOptions}
        handleFileDelete={handleFileDelete}
        handleRetry={handleRetry}
        handleCopy={handleCopy}
      />
    );
    const result = format(encodeHTMLEntities(userMessage));
    expect(result).not.toContain('<script>');
    expect(result).toContain('sanitized');

    const codeElement = screen.getByText('formatted');
    expect(codeElement.tagName).toBe('CODE');
    expect(codeElement).toHaveClass('inline-code');
  });

  it('renders file thumbnails when fileSrc is provided', () => {
    render(
      <ChatMessage
        isNew={false}
        message={message}
        lastIndex={lastIndex}
        loading={loading}
        fileSrc={fileSrc}
        selectedAssistant={selectedAssistant}
        handleAssistantChange={handleAssistantChange}
        assistantOptions={assistantOptions}
        handleFileDelete={handleFileDelete}
        handleRetry={handleRetry}
        handleCopy={handleCopy}
      />
    );
    expect(
      screen.getByTitle('Thumbnail for uploaded file 1')
    ).toBeInTheDocument();
    expect(
      screen.getByTitle('Thumbnail for uploaded file 2')
    ).toBeInTheDocument();
  });

  it('does not render image thumbnails for new messages', () => {
    render(
      <ChatMessage
        isNew={true}
        message={message}
        lastIndex={lastIndex}
        loading={loading}
        fileSrc={fileSrc}
        selectedAssistant={selectedAssistant}
        handleAssistantChange={handleAssistantChange}
        assistantOptions={assistantOptions}
        handleFileDelete={handleFileDelete}
        handleRetry={handleRetry}
        handleCopy={handleCopy}
      />
    );
    expect(
      screen.queryByTitle('Uploaded image thumbnail 1')
    ).not.toBeInTheDocument();
    expect(
      screen.queryByTitle('Uploaded image thumbnail 2')
    ).not.toBeInTheDocument();
  });

  it('renders FileListWithModal when fileSrc is not empty', () => {
    render(
      <ChatMessage
        isNew={false}
        message={message}
        lastIndex={true}
        loading={false}
        fileSrc={fileSrc}
        selectedAssistant={selectedAssistant}
        handleAssistantChange={handleAssistantChange}
        assistantOptions={assistantOptions}
        handleFileDelete={handleFileDelete}
        handleRetry={handleRetry}
        handleCopy={handleCopy}
      />
    );
    expect(
      screen.getByTitle('Thumbnail for uploaded file 1')
    ).toBeInTheDocument();
    expect(
      screen.getByTitle('Thumbnail for uploaded file 2')
    ).toBeInTheDocument();
  });

  it('does not render FileListWithModal when fileSrc is empty', () => {
    render(
      <ChatMessage
        isNew={false}
        message={message}
        lastIndex={true}
        loading={false}
        fileSrc={[]}
        selectedAssistant={selectedAssistant}
        handleAssistantChange={handleAssistantChange}
        assistantOptions={assistantOptions}
        handleFileDelete={handleFileDelete}
        handleRetry={handleRetry}
        handleCopy={handleCopy}
      />
    );
    expect(
      screen.queryByTitle('Thumbnail for uploaded file 1')
    ).not.toBeInTheDocument();
  });

  it('renders code blocks in ai answer correctly', () => {
    const messageWithCode = {
      question: 'Give me a code example.',
      answer:
        'Here\'s a code example:\n<pre>><code class="python">print("Hello, World!")\n</code></pre>',
      assistant: 'gpt-4'
    };
    render(
      <ChatMessage
        isNew={false}
        message={messageWithCode}
        lastIndex={lastIndex}
        loading={loading}
        fileSrc={fileSrc}
        selectedAssistant={selectedAssistant}
        handleAssistantChange={handleAssistantChange}
        assistantOptions={assistantOptions}
        handleFileDelete={handleFileDelete}
        handleRetry={handleRetry}
        handleCopy={handleCopy}
      />
    );
    expect(screen.getByText("Here's a code example:")).toBeInTheDocument();
    expect(screen.getByText('print("Hello, World!")')).toBeInTheDocument();
    const codeBlock = screen.getByText('print("Hello, World!")');
    expect(codeBlock.closest('pre')).toBeInTheDocument();
    expect(codeBlock.closest('code')).toHaveClass('python');
  });

  it('renders retry and copy buttons when lastIndex is true', () => {
    render(
      <ChatMessage
        isNew={false}
        message={message}
        lastIndex={true}
        loading={false}
        fileSrc={fileSrc}
        selectedAssistant={selectedAssistant}
        handleAssistantChange={handleAssistantChange}
        assistantOptions={assistantOptions}
        handleFileDelete={handleFileDelete}
        handleRetry={handleRetry}
        handleCopy={handleCopy}
      />
    );
    expect(screen.getByText('Retry')).toBeInTheDocument();
    expect(screen.getByText('Copy')).toBeInTheDocument();
  });

  it('renders assistant selector with selected assistant when lastIndex is true', () => {
    render(
      <ChatMessage
        isNew={false}
        message={message}
        lastIndex={true}
        loading={false}
        fileSrc={fileSrc}
        selectedAssistant={selectedAssistant}
        handleAssistantChange={handleAssistantChange}
        assistantOptions={assistantOptions}
        handleFileDelete={handleFileDelete}
        handleRetry={handleRetry}
        handleCopy={handleCopy}
      />
    );
    // Check if the selected model text is present
    expect(screen.getByText('GPT-4 Assistant')).toBeInTheDocument();

    // Check if the model selector button is present
    const modelSelectorButton = screen.getByLabelText('Change Assistant');
    expect(modelSelectorButton).toBeInTheDocument();
  });

  it('calls handleRetry when retry button is clicked', () => {
    render(
      <ChatMessage
        isNew={false}
        message={message}
        lastIndex={true}
        loading={false}
        fileSrc={fileSrc}
        selectedAssistant={selectedAssistant}
        handleAssistantChange={handleAssistantChange}
        assistantOptions={assistantOptions}
        handleFileDelete={handleFileDelete}
        handleRetry={handleRetry}
        handleCopy={handleCopy}
      />
    );
    const retryButton = screen.getByText('Retry');
    fireEvent.click(retryButton);
    expect(handleRetry).toHaveBeenCalledTimes(1);
  });

  it('calls handleCopy when copy button is clicked', () => {
    render(
      <ChatMessage
        isNew={false}
        message={message}
        lastIndex={true}
        loading={false}
        fileSrc={fileSrc}
        selectedAssistant={selectedAssistant}
        handleAssistantChange={handleAssistantChange}
        assistantOptions={assistantOptions}
        handleFileDelete={handleFileDelete}
        handleRetry={handleRetry}
        handleCopy={handleCopy}
      />
    );
    const copyButton = screen.getByText('Copy');
    fireEvent.click(copyButton);
    expect(handleCopy).toHaveBeenCalledTimes(1);
  });

  it('calls handleAssistantChange when a new model is selected', () => {
    render(
      <ChatMessage
        isNew={false}
        message={message}
        lastIndex={true}
        loading={false}
        fileSrc={fileSrc}
        selectedAssistant={selectedAssistant}
        handleAssistantChange={handleAssistantChange}
        assistantOptions={assistantOptions}
        handleFileDelete={handleFileDelete}
        handleRetry={handleRetry}
        handleCopy={handleCopy}
      />
    );

    const customSelectTrigger = screen.getByText('GPT-4 Assistant');

    // Simulate clicking the select trigger to open the dropdown
    fireEvent.click(customSelectTrigger);

    const gpt35Option = screen.getByText('Default GPT-3.5 Turbo');
    fireEvent.click(gpt35Option);

    expect(handleAssistantChange).toHaveBeenCalledWith(assistantOptions[0]);
  });

  it('does not render retry and copy buttons when lastIndex is false', () => {
    render(
      <ChatMessage
        isNew={false}
        message={message}
        lastIndex={false}
        loading={false}
        fileSrc={fileSrc}
        selectedAssistant={selectedAssistant}
        handleAssistantChange={handleAssistantChange}
        assistantOptions={assistantOptions}
        handleFileDelete={handleFileDelete}
        handleRetry={handleRetry}
        handleCopy={handleCopy}
      />
    );
    expect(screen.queryByText('Retry')).not.toBeInTheDocument();
    expect(screen.queryByText('Copy')).not.toBeInTheDocument();
  });

  it('ChatMessage component snapshot', () => {
    const { asFragment } = render(
      <ChatMessage
        isNew={false}
        message={message}
        lastIndex={lastIndex}
        loading={loading}
        fileSrc={fileSrc}
        selectedAssistant={selectedAssistant}
        handleAssistantChange={handleAssistantChange}
        assistantOptions={assistantOptions}
        handleFileDelete={handleFileDelete}
        handleRetry={handleRetry}
        handleCopy={handleCopy}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
