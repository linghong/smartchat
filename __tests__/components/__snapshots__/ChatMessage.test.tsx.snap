// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatMessage Component ChatMessage component snapshot 1`] = `
<DocumentFragment>
  <article
    aria-label="user-message"
    class="flex px-3 py-2 text-black"
  >
    <div
      class="w-16 px-2 flex flex-col justify-start items-center user"
    >
      <img
        alt="User avatar"
        class="h-8 w-8 mr-4 rounded-sm"
        data-nimg="1"
        decoding="async"
        fetchpriority="high"
        height="30"
        src="/_next/image?url=%2Fuser.png&w=64&q=75"
        srcset="/_next/image?url=%2Fuser.png&w=32&q=75 1x, /_next/image?url=%2Fuser.png&w=64&q=75 2x"
        style="color: transparent;"
        width="30"
      />
    </div>
    <div
      class="flex-1 user-question"
    >
      <div
        class="user-question whitespace-pre-wrap"
      >
        What is AI?
      </div>
      <div
        class="flex flex-row justify-start w-full p-2"
      >
        <div
          class="w-20 h-full mx-2 relative"
        >
          <div
            aria-label="Click to view larger file 1"
            class="mr-2 p-1 bg-clip-border bg-slate-100 border-2 border-violet-100"
            role="button"
            tabindex="0"
            title="Thumbnail for uploaded file 1"
          >
            <div
              class="bg-no-repeat bg-center bg-contain min-h-[60px] h-full w-full cursor-pointer"
              style="background-image: url(/path/to/image1.png);"
            />
          </div>
        </div>
        <div
          class="w-20 h-full mx-2 relative"
        >
          <div
            aria-label="Click to view larger file 2"
            class="mr-2 p-1 bg-clip-border bg-slate-100 border-2 border-violet-100"
            role="button"
            tabindex="0"
            title="Thumbnail for uploaded file 2"
          >
            <div
              class="bg-no-repeat bg-center bg-contain min-h-[60px] h-full w-full cursor-pointer"
              style="background-image: url(path/to/image2.png);"
            />
          </div>
        </div>
      </div>
    </div>
  </article>
  <article
    aria-labelledby="ai-response"
    class="flex px-3 py-2 bg-slate-100 text-black ease-in duration-300"
  >
    <div
      class="w-16 px-2 flex flex-col justify-start items-center ai"
    >
      <img
        alt="AI avatar"
        class="h-8 w-8 mr-4 rounded-sm false"
        data-nimg="1"
        decoding="async"
        fetchpriority="high"
        height="30"
        src="/_next/image?url=%2Fbot.png&w=64&q=75"
        srcset="/_next/image?url=%2Fbot.png&w=32&q=75 1x, /_next/image?url=%2Fbot.png&w=64&q=75 2x"
        style="color: transparent;"
        width="30"
      />
      <label
        class="text-xs"
      >
        gpt-4
      </label>
    </div>
    <div
      class="flex-1 ai-answer space-wrap"
    >
      <span>
        <p>
          AI stands for Artificial Intelligence.
        </p>
        

      </span>
      <div
        class="flex justify-between items-center text-xs"
      >
        <div
          class="flex space-x-1"
        >
          <button
            aria-label="Retry AI response"
            class="justify-center whitespace-nowrap text-sm font-medium ring-offset-white focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 hover:text-gray-900 dark:hover:bg-gray-800 dark:hover:text-gray-50 h-9 flex items-center p-1 text-gray-500 rounded-lg hover:bg-gray-200 transition-colors duration-200"
          >
            <svg
              class="lucide lucide-refresh-cw w-4 h-4 mr-1"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M3 12a9 9 0 0 1 9-9 9.75 9.75 0 0 1 6.74 2.74L21 8"
              />
              <path
                d="M21 3v5h-5"
              />
              <path
                d="M21 12a9 9 0 0 1-9 9 9.75 9.75 0 0 1-6.74-2.74L3 16"
              />
              <path
                d="M8 16H3v5"
              />
            </svg>
            <span
              class="text-xs"
            >
              Retry
            </span>
          </button>
          <button
            aria-label="Copy AI response"
            class="justify-center whitespace-nowrap text-sm font-medium ring-offset-white focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 hover:text-gray-900 dark:hover:bg-gray-800 dark:hover:text-gray-50 h-9 flex items-center p-1 text-gray-500 rounded-lg hover:bg-gray-200 transition-colors duration-200"
          >
            <svg
              class="lucide lucide-copy w-4 h-4 mr-1"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="14"
                rx="2"
                ry="2"
                width="14"
                x="8"
                y="8"
              />
              <path
                d="M4 16c-1.1 0-2-.9-2-2V4c0-1.1.9-2 2-2h10c1.1 0 2 .9 2 2"
              />
            </svg>
            <span
              class="text-xs"
            >
              Copy
            </span>
          </button>
        </div>
        <div
          class="flex items-center space-x-2"
        >
          <button
            aria-label="Change Assistant"
            class="justify-center whitespace-nowrap text-sm font-medium ring-offset-white focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 hover:text-gray-900 dark:hover:bg-gray-800 dark:hover:text-gray-50 h-10 flex items-center p-1 text-gray-500 rounded-lg hover:bg-gray-200 transition-colors duration-200"
          >
            <svg
              class="lucide lucide-wand-sparkles w-4 h-4 mr-1"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m21.64 3.64-1.28-1.28a1.21 1.21 0 0 0-1.72 0L2.36 18.64a1.21 1.21 0 0 0 0 1.72l1.28 1.28a1.2 1.2 0 0 0 1.72 0L21.64 5.36a1.2 1.2 0 0 0 0-1.72"
              />
              <path
                d="m14 7 3 3"
              />
              <path
                d="M5 6v4"
              />
              <path
                d="M19 14v4"
              />
              <path
                d="M10 2v2"
              />
              <path
                d="M7 8H3"
              />
              <path
                d="M21 16h-4"
              />
              <path
                d="M11 3H9"
              />
            </svg>
          </button>
          <div
            class="w-72"
          >
            <div
              class="relative"
              id="assistant"
            >
              <div
                aria-controls="options"
                aria-expanded="false"
                class="select-trigger w-full p-2 border border-gray-300 rounded-md cursor-pointer bg-gray-50 text-left text-xs py-2"
                role="combobox"
              >
                <span
                  class="text-gray-900"
                >
                  GPT-4 Assistant
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </article>
</DocumentFragment>
`;
